### ISSUE TITLE
Recursion not supported error in non-recursive code and Julia crashes when inspecting

### ISSUE THREAD

**Issue by mohamed82008** [JuliaGPU/CUDAnative.jl#292]
The following code errors with
```julia
ERROR: GPU compilation of kernel(CuDeviceArray{SArray{Tuple{3},Float64,1,3},1,CUDAnative.AS.Global}, CuDeviceArray{SArray{Tuple{3,3},Float64,2,9},1,CUDAnative.AS.Global}, CuDeviceArray{Float64,1,CUDAnative.AS.Global}, CuDeviceArray{Int64,1,CUDAnative.AS.Global}, Int64) failed
KernelError: recursion is currently not supported

Try inspecting the generated code with any of the @device_code_... macros.

Stacktrace:
 [1] #IOBuffer#300 at iobuffer.jl:112
 [2] Type at none:0
 [3] print_to_string at strings/io.jl:112
 [4] #IOBuffer#299 at iobuffer.jl:91
 [5] #IOBuffer#300 at iobuffer.jl:112
 [6] print_to_string at strings/io.jl:112
 [7] throw_overflowerr_binaryop at checked.jl:157
 [8] _unsafe_getindex at multidimensional.jl:600
 [9] kernel at REPL[14]:2
Stacktrace:
 [1] (::getfield(CUDAnative, Symbol("#hook_emit_function#58")){CUDAnative.CompilerContext,Array{Core.MethodInstance,1}})(::Core.MethodInstance, ::Core.CodeInfo, ::UInt64) at C:\Users\user\.julia\packages\CUDAnative\nqSUm\src\compiler\irgen.jl:97
 [2] irgen(::CUDAnative.CompilerContext) at C:\Users\user\.julia\packages\CUDAnative\nqSUm\src\compiler\irgen.jl:133
 [3] #compile#80(::Bool, ::Function, ::CUDAnative.CompilerContext) at .\logging.jl:308
 [4] compile at C:\Users\user\.julia\packages\CUDAnative\nqSUm\src\compiler\driver.jl:49 [inlined]
 [5] #compile#79(::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}, ::Function, ::CuDevice, ::Any, ::Any) at C:\Users\user\.julia\packages\CUDAnative\nqSUm\src\compiler\driver.jl:28
 [6] macro expansion at C:\Users\user\.julia\packages\CUDAnative\nqSUm\src\compiler\driver.jl:16 [inlined]
 [7] #cufunction#91(::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}, ::typeof(cufunction), ::typeof(kernel), ::Type{Tuple{CuDeviceArray{SArray{Tuple{3},Float64,1,3},1,CUDAnative.AS.Global},CuDeviceArray{SArray{Tuple{3,3},Float64,2,9},1,CUDAnative.AS.Global},CuDeviceArray{Float64,1,CUDAnative.AS.Global},CuDeviceArray{Int64,1,CUDAnative.AS.Global},Int64}}) at C:\Users\user\.julia\packages\CUDAnative\nqSUm\src\execution.jl:234
 [8] cufunction(::Function, ::Type) at C:\Users\user\.julia\packages\CUDAnative\nqSUm\src\execution.jl:234
 [9] top-level scope at C:\Users\user\.julia\packages\CUDAnative\nqSUm\src\execution.jl:202
 [10] top-level scope at gcutils.jl:87
 [11] top-level scope at C:\Users\user\.julia\packages\CUDAnative\nqSUm\src\execution.jl:199
```
and I don't understand why. `@device_code_lowered` crashes Julia.

```julia
using CUDAnative, CUDAdrv, CuArrays, StaticArrays

k = 3
n = 10

Ks = [SMatrix{k, k, Float64}(rand(k,k)) for i in 1:n] |> CuArrays.CuArray
fs = [SVector{k, Float64}(rand(k)) for i in 1:n] |> CuArrays.CuArray
newfs = similar(fs)
m = fill(1.0, n) |> CuArrays.CuArray
inds = rand(1:n, n) |> CuArrays.CuArray

dev = CUDAdrv.device()
ctx = CUDAdrv.CuContext(dev)
MAX_THREADS_PER_BLOCK = CUDAdrv.attribute(dev, CUDAdrv.MAX_THREADS_PER_BLOCK)
threads = min(n, MAX_THREADS_PER_BLOCK)
blocks = ceil.(Int, n / threads)

function kernel(fs::AbstractVector{TV}, Ks, m, inds, n) where {N, T, TV<:SVector{N,T}}
    i = (blockIdx().x-1) * blockDim().x + threadIdx().x
    if i <= n
        p = m[inds[i]]
        f = fs[i]
        for j in 1:N
    	    f = SVector{N, T}(f[1:j-1]..., 1.0, f[j+1:end]...)
    	end
	fs[i] = SVector{1,T}((p,)) .* (Ks[i] * f)
    end
   return
end

@cuda blocks=blocks threads=threads kernel(fs,  Ks, m, inds, n)
```

```julia
julia> versioninfo()
Julia Version 1.0.1
Commit 0d713926f8 (2018-09-29 19:05 UTC)
Platform Info:
  OS: Windows (x86_64-w64-mingw32)
  CPU: Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-6.0.0 (ORCJIT, skylake)
Environment:
  JULIA_HOME_v0.6 = C:\Users\user\AppData\Local\Julia-0.6.2\bin
  JULIA_HOME_v1.0 = C:\Users\user\AppData\Local\Julia-1.0.0\bin
```

This is what I get with `@device_code_lowered kernel(fs,  Ks, m, inds, n)`.

```julia
Please submit a bug report with steps to reproduce this fault, and any error messages that follow (in their entirety). Thanks.
Exception: EXCEPTION_ACCESS_VIOLATION at 0x5cf3c8914 -- _ZN4llvm16DAGTypeLegalizer3runEv at C:\Users\user\AppData\Local\Julia-1.0.1\bin\LLVM.dll (unknown line)
in expression starting at no file:0
_ZN4llvm16DAGTypeLegalizer3runEv at C:\Users\user\AppData\Local\Julia-1.0.1\bin\LLVM.dll (unknown line)
_ZN4llvm12SelectionDAG13LegalizeTypesEv at C:\Users\user\AppData\Local\Julia-1.0.1\bin\LLVM.dll (unknown line)
_ZN4llvm16SelectionDAGISel17CodeGenAndEmitDAGEv at C:\Users\user\AppData\Local\Julia-1.0.1\bin\LLVM.dll (unknown line)
_ZN4llvm16SelectionDAGISel20SelectAllBasicBlocksERKNS_8FunctionE at C:\Users\user\AppData\Local\Julia-1.0.1\bin\LLVM.dll (unknown line)
_ZN4llvm16SelectionDAGISel20runOnMachineFunctionERNS_15MachineFunctionE at C:\Users\user\AppData\Local\Julia-1.0.1\bin\LLVM.dll (unknown line)
_ZN12_GLOBAL__N_115X86DAGToDAGISel20runOnMachineFunctionERN4llvm15MachineFunctionE at C:\Users\user\AppData\Local\Julia-1.0.1\bin\LLVM.dll (unknown line)
_ZN4llvm19MachineFunctionPass13runOnFunctionERNS_8FunctionE at C:\Users\user\AppData\Local\Julia-1.0.1\bin\LLVM.dll (unknown line)
_ZN4llvm13FPPassManager13runOnFunctionERNS_8FunctionE at C:\Users\user\AppData\Local\Julia-1.0.1\bin\LLVM.dll (unknown line)
_ZN4llvm13FPPassManager11runOnModuleERNS_6ModuleE at C:\Users\user\AppData\Local\Julia-1.0.1\bin\LLVM.dll (unknown line)
_ZN4llvm6legacy15PassManagerImpl3runERNS_6ModuleE at C:\Users\user\AppData\Local\Julia-1.0.1\bin\LLVM.dll (unknown line)
operator() at /home/Administrator/buildbot/worker/package_win64/build/src/home/Administrator/buildbot/worker/package_win64/build/src\jitlayers.cpp:481
addModule at /home/Administrator/buildbot/worker/package_win64/build/src/home/Administrator/buildbot/worker/package_win64/build/usr/include/llvm/ExecutionEngine/Orc\IRCompileLayer.h:57 [inlined]
addModule at /home/Administrator/buildbot/worker/package_win64/build/src/home/Administrator/buildbot/worker/package_win64/build/src\jitlayers.cpp:608
jl_add_to_ee at /home/Administrator/buildbot/worker/package_win64/build/src/home/Administrator/buildbot/worker/package_win64/build/src\jitlayers.cpp:855 [inlined]
jl_finalize_function at /home/Administrator/buildbot/worker/package_win64/build/src/home/Administrator/buildbot/worker/package_win64/build/src\jitlayers.cpp:863
getAddressForFunction at /home/Administrator/buildbot/worker/package_win64/build/src/home/Administrator/buildbot/worker/package_win64/build/src\codegen.cpp:1299
jl_generate_fptr at /home/Administrator/buildbot/worker/package_win64/build/src/home/Administrator/buildbot/worker/package_win64/build/src\codegen.cpp:1410
jl_compile_method_internal at /home/Administrator/buildbot/worker/package_win64/build/src\gf.c:1823 [inlined]
jl_fptr_trampoline at /home/Administrator/buildbot/worker/package_win64/build/src\gf.c:1830
jl_apply_generic at /home/Administrator/buildbot/worker/package_win64/build/src\gf.c:2184
top-level scope at C:\Users\user\.julia\packages\CUDAnative\nqSUm\src\reflection.jl:155 [inlined]
top-level scope at .\none:0
jl_fptr_trampoline at /home/Administrator/buildbot/worker/package_win64/build/src\gf.c:1831
jl_toplevel_eval_flex at /home/Administrator/buildbot/worker/package_win64/build/src\toplevel.c:807
jl_toplevel_eval_in at /home/Administrator/buildbot/worker/package_win64/build/src\builtins.c:622
eval at .\boot.jl:319
jl_apply_generic at /home/Administrator/buildbot/worker/package_win64/build/src\gf.c:2184
eval_user_input at C:\cygwin\home\Administrator\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.0\REPL\src\REPL.jl:85
macro expansion at C:\cygwin\home\Administrator\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.0\REPL\src\REPL.jl:117 [inlined]
#28 at .\task.jl:259
jl_apply_generic at /home/Administrator/buildbot/worker/package_win64/build/src\gf.c:2184
jl_apply at /home/Administrator/buildbot/worker/package_win64/build/src\julia.h:1537 [inlined]
start_task at /home/Administrator/buildbot/worker/package_win64/build/src\task.c:268
Allocations: 47728503 (Pool: 47720323; Big: 8180); GC: 125

Please submit a bug report with steps to reproduce this fault, and any error messages that follow (in their entirety). Thanks.
Exception: EXCEPTION_ACCESS_VIOLATION at 0xffffffff00000020 -- unknown function (ip: FFFFFFFF00000020)
in expression starting at no file:0
unknown function (ip: FFFFFFFF0000001F)
Allocations: 47734449 (Pool: 47726269; Big: 8180); GC: 125
```

**Comment by maleadt** [comment]
https://github.com/JuliaGPU/CUDAnative.jl/issues/278
Fixed in Julia 1.0.2

**Comment by maleadt** [comment]
The EXCEPTION_ACCESS_VIOLATION is strange though. But it traces to Julia code so nothing we can do about that.
